Add cypress E2E tests

brentmclark opened the pull request:
I didn't want to take this PR too far until I had buy-in from the team that `cypress` was OK to implement.  For starters, I've done the following:

- [x] test each filter with a single test to ensure the baseline functionality works as expected.
- [x] Hook up scripts for running locally and in CI.
- [x] Allow the `data-testid` prop to be passed down in cases where it was not allowed to do so.
- [x] Remove any unnecessary files
- [x] Update the `README` with new procedures for testing locally

I expect this to turn into a discussion well before it gets committed, so please don't hesitate to suggest changes, ask for clarification, or anything else that makes you feel comfortable with the proposed approach.

moshfeu commented:
Thank you! We didn't think about e2e's because it's a temporary project but any effort to improve the quality is good! The current tests (almost of them) now are for validate the mentors schema.
Anyway, personally I'm not familiar with cypress but it works, so it's good :) If there are non necessary files, you can delete them.
I like to hear more opinions.

brentmclark commented:
> Thank you! We didn't think about e2e's because it's a temporary project but any effort to improve the quality is good!

It didn't take too long to get this all up and running; 2 hours or so, and it was fun!  My work does not define me, so don't hesitate to close the PR if now isn't the right time for `cypress`.  I appreciate the fact that you're even considering it 😃

moshfeu commented:
Sine there are no objections I'll merge it and will re-think before we wll move to full app.
Can you just remove the unnecessary (empty) files?

brentmclark commented:
Awesome news!  I don't see any empty files, but there's 2 or 3 that I might be able to remove as long as they aren't required by cypress.  I'll give it a think, and also touch up the `README` before I check all the boxes.  Once we're at 5/5 it should be good to go.

moshfeu commented:
By empty I ment unnecessary files (all the code in comments and so).
Thanks

tureey commented:
From my point of view it looks great.

BUT, I would create commands to do some actions like, filterByName, SelectCountry and things like this.

What is the advantadge of that? Is a function which wraps all the related thing (select the DOM and do the action).

**pseudocode-here**
```
describe('Mentor Filtering', () => {
  it('can filter by technology', () => {
      cy.visit('/')
      .filterByName('Brent M Clark')
      .filterBySkill('reactjs')
      
      cy.getByTestId('mentor-card').should('have.length', 1);
  });
});
```

where `filterByName` and `filterBySkill` are [commands](https://docs.cypress.io/api/cypress-api/custom-commands.html).
The idea behind commands is to encapsulate code that is going to be repeated, isolate test logic and improve performance.

[You can checkout this talk][(https://www.youtube.com/watch?v=7N63cMKosIE)

brentmclark commented:
Apologies for the delay; I had a busy weekend.  

### Regarding Cypress

I've incorporated *most* of the requested feedback, but only added a cypress command for filtering by name.  All other user behaviors did not meet my normal criteria for abstraction (rule of 3), so I chose to avoid being too hasty this time around.  *If* we end up making greater use of cypress, and *if* we perform more filtering tests, I will gladly add more commands.  That is, of course, unless someone tells me to "do it anyway" 😄 

### Regarding the CI changes

We were previously caching the `node_modules` folder instead of leveraging the `yarn-cache` which was just fine until `cypress` landed on the scene.  The post-install hook cypress uses to stand itself up does not play nicely with a cached `node_modules` directory, so I switched the cache to `yarn-cache`.

Additionally, I switched to using `yarn` wherever possible to maximize the yarn cache.

moshfeu commented:
Thanks for your great effort!

tureey commented:
@brentmclark amazing job!

Good explanation of the changes applies!

Congratulations :tada:

brentmclark commented:
Thanks @moshfeu and @tureey!

crysfel commented:
I've only heard great things about cypress! Definitely would like to give it a try, thanks!
